### PR TITLE
runtime: misc pool cleanup

### PR DESCRIPTION
**Pull request by carllerche**
- Remove builders for internal types
- Avoid duplicating the blocking pool when using the concurrent
  scheduler.
- misc smaller cleanup

### CODE REVIEW

File: tokio/src/runtime/builder.rs [hunk 1]
```diff
@@ -75,6 +71,12 @@ enum Kind {
     ThreadPool,
 }
 
+#[cfg(not(loom))]
+type Callback = Arc<dyn Fn() + Send + Sync>;
+
+#[cfg(loom)]
+type Callback = Arc<Box<dyn Fn() + Send + Sync>>;
```
Comment (hawkw) [inline tokio/src/runtime/builder.rs]: Why is the box required with `loom`?

Comment (carllerche) [reply tokio/src/runtime/builder.rs]: Loom doesn't support trait objs... partially because it isn't really doable on stable.

Comment (hawkw) [reply tokio/src/runtime/builder.rs]: aha

File: tokio/src/runtime/builder.rs [hunk 2]
```diff
@@ -351,48 +355,52 @@ impl Builder {
...
+                    next();
+
+                    // Call the after call back
+                    if let Some(before_stop) = before_stop.as_ref() {
+                        before_stop();
+                    }
                 })
-                .build(move |index| timers[index].lock().unwrap().take().unwrap())
+            })
+                as Box<dyn Fn(usize, &mut dyn FnMut()) + Send + Sync>);
```
Comment (hawkw) [inline tokio/src/runtime/builder.rs]: nit: this formatting feels a little weird to me...what about using a separate let binding for the closure? maybe that wouldn't be any better.

File: tokio/src/runtime/builder.rs [hunk 3]
```diff
@@ -351,48 +355,52 @@ impl Builder {
...
-
-                        // Run the worker
-                        next();
-
-                        // Call the after call back
-                        if let Some(before_stop) = before_stop.as_ref() {
-                            before_stop();
-                        }
-                    })
+            let around_worker = Arc::new(Box::new(move |index, next: &mut dyn FnMut()| {
```
Comment (jonhoo) [inline tokio/src/runtime/builder.rs]: Do we want to avoid the extra `Box` in `not(loom)` here too?

File: tokio/src/runtime/thread_pool/mod.rs [hunk 4]
```diff
@@ -44,3 +38,201 @@ const LOCAL_QUEUE_CAPACITY: usize = 256;
...
+        // ThreadPool::from_parts(spawner, shutdown_rx, blocking)
+        ThreadPool {
+            spawner,
+            shutdown_rx,
+            blocking,
+        }
+    }
+
+    /*
+    fn from_parts(
```
Comment (jonhoo) [inline tokio/src/runtime/thread_pool/mod.rs]: Was this included intentionally?

File: tokio/src/runtime/builder.rs (hunk 3 above)
Comment (carllerche) [reply tokio/src/runtime/builder.rs]: I will probably take a stab at improving this in a future PR.

File: tokio/src/runtime/builder.rs (hunk 2 above)
Comment (carllerche) [reply tokio/src/runtime/builder.rs]: open issue w/ rustfmt / clippy :P

Comment (carllerche) [reply tokio/src/runtime/builder.rs]: I'm guessing this specific code won't stay around in follow up PRs, so i'm going to skip the nit.

**Pull request merged**